### PR TITLE
Facilitate template update [WIP]

### LINKED ISSUE #61: How to update existing birds with latest from Cookiecutter template?
## Description

We consistently improve the template while we are using the WPS birds. What could be ways to update existing birds with the improvements?

## Environment

* Cookiecutter version used, if any:
* Python version, if any:
* Operating System:

### PR DESCRIPTION
**Pull request by huard**
* Add update_from_cc.py script 
* Replace jinja timestamps by cookiecutter._timestamp

TODO: 
* Add instructions
* Test

Fixes #61

### CODE REVIEW

File: {{cookiecutter.project_repo_name}}/update_from_cc.py [hunk 1]
```diff
@@ -1,21 +1,52 @@
+"""
+This script replays the cookiecutter repository creation with the original values.
+
+The intended use is to inject template changes into an existing repo.
+
+How-to
+======
+* Make sure that there is no uncommited files.
```
Comment (Zeitsperre) [inline {{cookiecutter.project_repo_name}}/update_from_cc.py]:
```suggestion
* Make sure that there are no files left uncommitted.
```

**Comment by huard** [comment]
Maybe another option would be https://scaraplate.readthedocs.io/en/latest/

**Comment by cehbrecht** [comment]
@huard looks good. Thanks for adding this. Feel free to merge.

**Comment by cehbrecht** [comment]
@huard is it ready to merge?

**Comment by huard** [comment]
I was waiting for some feedback by @davidcaron who planned to test it with Finch.

**Comment by davidcaron** [comment]
Yes, I'll try it right now.

File: {{cookiecutter.project_repo_name}}/update_from_cc.py [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+* Fix the conflicts and merge
+
+.. code:: bash
+
+    git checkout -b orig_cc_update <SHA>
+    git clean -fxd
+    cd ..
+    python {{ cookiecutter.project_slug }}/update_from_cc.py
+    cd {{ cookiecutter.project_slug }}
+    git add * */.*
```
Comment (davidcaron) [inline {{cookiecutter.project_repo_name}}/update_from_cc.py]: I think there is a typo here, I used `git add * **/.*` but is `git add .` equivalent?

Comment (huard) [reply {{cookiecutter.project_repo_name}}/update_from_cc.py]: On my end `git add .` did not add dot files.

**Comment by huard** [comment]
@davidcaron Would you recommend defining a `template` branch that would track the cookiecutter template ?

File: {{cookiecutter.project_repo_name}}/update_from_cc.py (hunk 2 above)
Comment (davidcaron) [reply {{cookiecutter.project_repo_name}}/update_from_cc.py]: Good to know

**Comment by davidcaron** [comment]
> @davidcaron Would you recommend defining a `template` branch that would track the cookiecutter template ?

I think this is a good idea, but when I get some time, I'll check out scaraplate like you mentioned or https://github.com/timothycrosley/cruft/

**Comment by cehbrecht** [comment]
I have used cruft to update hummingbird ... and it worked. There some merge conflicts I had to solve:
https://github.com/bird-house/hummingbird/pull/35

**Comment by Zeitsperre** [comment]
@cehbrecht I just discovered cruft the other day. Is it efficient/easy to work with?

**Comment by cehbrecht** [comment]
I find `cruft` easy to work with. I have tried now also [scaraplate](https://scaraplate.readthedocs.io/en/latest/template.html). That one looks more advanced. Currently I would favor `cruft` ... just a feeling.

@huard @davidcaron Would you like to update this PR to use one of these tools?

**Comment by huard** [comment]
Yes, it looks easier that what I came up with. I suggest we open a new similar PR using cruft.

**Comment by tlvu** [comment]
Found another cookiecutter updater: https://pypi.org/project/cookiecutter-project-upgrader/.

Clearly keeping projects up-to-date with the original cookiecutter template is a painful and popular topic.

Not clear which of the 3 tools we should go with yet, but here are my selection criteria:
* simple to setup and use
* handle merge conflict well for divergence between the original template and the instantiated template

We could help the merge conflict part by refactoring the template to split generic "lines" out to separate file(s).  By having a clear physical boundary, we will be encouraged to write more generic "implementation" which will prevent merge conflicts down the lines.

**Comment by huard** [comment]
@tlvu This PR can probably be closed and redone from scratch using one package.

**Comment by tlvu** [comment]
Close this PR unmerged since it has been replaced by PR https://github.com/bird-house/cookiecutter-birdhouse/pull/85

**Pull request closed without merging**